patches/server/1051-Registry-Modification-API.patch

```diff
@@ -0,0 +1,966 @@
++        final Registry<M> registry,
++        final ResourceKey<M> key,
++        final M nms,
++        final RegistrationInfo registrationInfo,
++        final RegisterMethod<M, R> registerMethod
++    ) {
++        Preconditions.checkState(LaunchEntryPointHandler.INSTANCE.hasEntered(Entrypoint.BOOTSTRAPPER), registry.key() + " tried to run modification listeners before bootstrappers have been called"); // verify that bootstrappers have been called
++        final @Nullable RegistryEntryInfo<M, T> entry = PaperRegistries.getEntry(registry.key());
++        if (!(entry instanceof RegistryEntry.Modifiable<?, ?, ?>) || !this.additionHooks.hasHooks(entry.apiKey())) {
++            return registerMethod.register((WritableRegistry<M>) registry, key, nms, registrationInfo);
++        }
++        final RegistryEntry.Modifiable<M, T, B> modifiableEntry = (RegistryEntry.Modifiable<M, T, B>) entry;
++        final CraftRegistry<T, M> craftRegistry = (CraftRegistry<T, M>) PaperRegistryAccess.instance().getRegistry(entry.apiKey());
++        @SuppressWarnings("PatternValidation") final TypedKey<T> typedKey = TypedKey.create(entry.apiKey(), Key.key(key.location().getNamespace(), key.location().getPath()));
++        final B builder = modifiableEntry.fillBuilder(typedKey, nms);
++        return this.registerWithListeners(registry, craftRegistry.view, modifiableEntry, key, nms, builder, registrationInfo, registerMethod);
++    }
++
++    public <M, T extends org.bukkit.Keyed, B extends PaperRegistryBuilder<M, T>> void registerWithListeners(
++        final Registry<M> registry,
++        final RegistryView<T> registryView,
++        final RegistryEntry.Modifiable<M, T, B> entry,
++        final ResourceKey<M> key,
++        final @Nullable M oldNms,
++        final B builder,
++        final RegistrationInfo registrationInfo
++    ) {
++        this.registerWithListeners(registry, registryView, entry, key, oldNms, builder, registrationInfo, WritableRegistry::register);
++    }
++
++    public <M, T extends org.bukkit.Keyed, B extends PaperRegistryBuilder<M, T>, R> R registerWithListeners(
++        final Registry<M> registry,
++        final RegistryView<T> registryView,
++        final RegistryEntry.Modifiable<M, T, B> entry,
++        final ResourceKey<M> key,
++        final @Nullable M oldNms,
++        final B builder,
++        final RegistrationInfo registrationInfo,
++        final RegisterMethod<M, R> registerMethod
++    ) {
++        @Subst("namespace:key") final ResourceLocation beingAdded = key.location();
++        @SuppressWarnings("PatternValidation") final TypedKey<T> typedKey = TypedKey.create(entry.apiKey(), Key.key(beingAdded.getNamespace(), beingAdded.getPath()));
++        final RegistryAdditionEventImpl<T, B> event = entry.createAdditionEvent(typedKey, builder, registryView);
++        LifecycleEventRunner.INSTANCE.callEvent(this.additionHooks.getHook(entry.apiKey()), event);
```

Comment:
After removing the addition handler from the test plugin (only listening to preFreeze), an NPE is thrown because `getHook` returns null.

```
[LifecycleEventRunner] Could not run 'RegistryKeyImpl[key=minecraft:game_event] / pre-freeze' lifecycle event handler from Paper-Test-Plugin v1.0.0-SNAPSHOT
java.lang.NullPointerException: Cannot invoke "io.papermc.paper.plugin.lifecycle.event.types.AbstractLifecycleEventType.forEachHandler(java.util.function.Consumer, java.util.function.Predicate)" because "lifecycleEventType" is null
	at io.papermc.paper.plugin.lifecycle.event.LifecycleEventRunner.callEvent(LifecycleEventRunner.java:63) ~[main/:?]
	at io.papermc.paper.plugin.lifecycle.event.LifecycleEventRunner.callEvent(LifecycleEventRunner.java:58) ~[main/:?]
	at io.papermc.paper.registry.PaperRegistryListenerManager.registerWithListeners(PaperRegistryListenerManager.java:127) ~[main/:?]
	at io.papermc.paper.registry.PaperRegistryListenerManager.registerWithListeners(PaperRegistryListenerManager.java:111) ~[main/:?]
	at io.papermc.paper.registry.WritableCraftRegistry.register(WritableCraftRegistry.java:46) ~[main/:?]
	at io.papermc.paper.registry.event.PaperWritableRegistry.register(PaperWritableRegistry.java:21) ~[main/:?]
	at test-plugin-1.0.0-SNAPSHOT.jar/io.papermc.testplugin.TestPluginBootstrap.lambda$bootstrap$1(TestPluginBootstrap.java:32) ~[test-plugin-1.0.0-SNAPSHOT.jar:?]
```

Reply:
Nice catch, fixed.